core/unit_test/TestDeviceAndThreads.py

```diff
@@ -18,22 +18,7 @@ def GetFlag(flag, *extra_args):
     return int(p.stdout)
 
 def GetNumThreads(max_threads):
```

Comment:
Also import at the top
`from typing import Iterator`

```suggestion
def GetNumThreads(max_threads: int) -> Iterator[int]:
    """
    Generate a sequence of thread counts up to the specified maximum.

    If the "hwloc_enabled" flag is set, only yields [1, 2, max_threads].
    Otherwise, yields [1, 2, 3, 4, 5, max_threads], stopping early if
    any value exceeds or equals `max_threads`.

    Parameters
    ----------
    max_threads : int
        The maximum number of threads to yield.

    Yields
    ------
    int
        Thread counts to be used (including `max_threads` as the final value).
    """
    candidates = [1, 2] if GetFlag("hwloc_enabled") else [1, 2, 3, 4, 5]
    
    for x in candidates:
        if x >= max_threads:
            break
        yield x

    yield max_threads
```

Reply:
What benefit do we get from using Iterator?  Does it exceed the cognitive load?

Reply:
>     If the "hwloc_enabled" flag is set, only yields [1, 2, max_threads].
>     Otherwise, yields [1, 2, 3, 4, 5, max_threads], stopping early if
>     any value exceeds or equals `max_threads`.

This assumes that max_treads exceeds 5 which is not always true.

Reply:
If you feel strongly about documenting these one-off test helpers (none of which has docstring), I would suggest we open a follow up PR to document.
In short, I am not so keen on doing it here.

Reply:
> What benefit do we get from using Iterator? Does it exceed the cognitive load?

As it yields not returns, it generates an iterator
It is used for typing, which improves the readability

Reply:
> > ```
> > If the "hwloc_enabled" flag is set, only yields [1, 2, max_threads].
> > Otherwise, yields [1, 2, 3, 4, 5, max_threads], stopping early if
> > any value exceeds or equals `max_threads`.
> > ```
> 
> This assumes that max_treads exceeds 5 which is not always true.

That is what I am saying here. May be unclear?

> stopping early if any value exceeds or equals `max_threads`.

Reply:
> > What benefit do we get from using Iterator? Does it exceed the cognitive load?
> 
> As it yields not returns, it generates an iterator

Why is it better than just yielding raw integers?

Reply:
> > > What benefit do we get from using Iterator? Does it exceed the cognitive load?
> > 
> > 
> > As it yields not returns, it generates an iterator
> 
> Why is it better than just yielding raw integers?

I mean generators are iterators. Thus, the return type should be `Iterator[int]`
https://stackoverflow.com/questions/2776829/difference-between-pythons-generators-and-iterators